onnxruntime/contrib_ops/rocm/bert/attention_impl.cu

```diff
@@ -152,15 +151,27 @@ Status QkvToContext(
 
   // apply softmax and store result P to scratch2: BxNxSxS*
   if (use_raw_attention_mask) {  // 2d, 3d or 4d attention mask
+    // Raw attention mask could be 2D (BxS) or 3D (BxSxS*) or 4D(Bx1xMxM), where M is the max sequence length.
+    auto* mask = mask_index;
+    int4 strides;
     const int mask_dimension = static_cast<int>(mask_index_dims.size());
-    const int max_sequence_length = mask_dimension == 4 ? static_cast<int>(mask_index_dims[3]) : 0;
+    if (mask_dimension == 2) {
+      strides = {all_sequence_length, 0, 0, 1};
+    } else if (mask_dimension == 3) {
+      strides = {sequence_length * all_sequence_length, 0, all_sequence_length, 1};
+    } else if (mask_dimension == 4) {
+      int max_sequence_length = mask_dimension == 4 ? static_cast<int>(mask_index_dims[3]) : 0;
+      strides = {max_sequence_length * max_sequence_length, max_sequence_length, max_sequence_length, 1};
+      // offset to skip past sequence part, so that we can index it with [batch_index, 0, sequence_index, token_index]
+      mask = mask + past_sequence_length * max_sequence_length;
```

Comment:
Does this logic only support batch_size == 1?

Reply:
The offset is very generic actually, you should adopt it, seriously.

The tensor size attirbute the logical shape of a tensor.
The tensor strides attribute the physical layout in the memory of a tensor.
The start of the buffer, defines how we view the tensor.

In this case, we are creating a sub-view to the original tensor. The sub-view starts at `original_tensor[0,0, past_sequence_length, 0]`, The memory is not changed, so the strides are still refer to the original one. The sub-view is effectively `original_tensor[:, :, past_sequence_length:past_sequence_length+sequence_length, :total_sequence_length]`